packages/zoltan2/core/src/algorithms/color/Zoltan2_AlgHybridD2.hpp

```diff
@@ -0,0 +1,1323 @@
+      Kokkos::parallel_for(boundary_size, KOKKOS_LAMBDA(const int& i){
+        verts_to_recolor_view(i) = -1;
+      });
+      Kokkos::View<lno_t*, device_type, Kokkos::MemoryTraits<Kokkos::Atomic>> verts_to_recolor_atomic = verts_to_recolor_view;
+
+      Kokkos::View<int[1], device_type> verts_to_recolor_size("verts to recolor size");
+      verts_to_recolor_size(0) = 0;
+      Kokkos::View<int[1], device_type, Kokkos::MemoryTraits<Kokkos::Atomic>> verts_to_recolor_size_atomic = verts_to_recolor_size;
+
+      Kokkos::View<lno_t*, device_type> verts_to_send_view("verts to send", boundary_size);
+      Kokkos::parallel_for(boundary_size, KOKKOS_LAMBDA(const int& i){
+        verts_to_send_view(i) = -1;
+      });
+      Kokkos::View<lno_t*, device_type, Kokkos::MemoryTraits<Kokkos::Atomic>> verts_to_send_atomic = verts_to_send_view;
+
+      Kokkos::View<size_t[1], device_type> verts_to_send_size("verts to send size");
+      verts_to_send_size(0) = 0;
+      Kokkos::View<size_t[1], device_type, Kokkos::MemoryTraits<Kokkos::Atomic>> verts_to_send_size_atomic = verts_to_send_size;
+      if(verbose) std::cout<<comm->getRank()<<": initializing the list of vertices to send\n";
+      Kokkos::parallel_for(n_local, KOKKOS_LAMBDA(const int& i){
+        for(offset_t j = dist_offsets_dev(i); j < dist_offsets_dev(i+1); j++){
+          if((size_t)dist_adjs_dev(j) >= n_local){
+            verts_to_send_atomic(verts_to_send_size_atomic(0)++) = i;
+            break;
+          }
+          bool found = false;
+          for(offset_t k = dist_offsets_dev(dist_adjs_dev(j)); k < dist_offsets_dev(dist_adjs_dev(j)+1); k++){
+            if((size_t)dist_adjs_dev(k) >= n_local){
+              verts_to_send_atomic(verts_to_send_size_atomic(0)++) = i;
+              found = true;
+              break;
+            }
+          }
+          if(found) break;
+        }
+      });
+      Kokkos::fence();
+      Kokkos::View<lno_t*, device_type> boundary_verts_dev("boundary vertices",verts_to_send_size(0));
+      typename Kokkos::View<lno_t*, device_type>::HostMirror boundary_verts_host = Kokkos::create_mirror(boundary_verts_dev);
+      Kokkos::deep_copy(boundary_verts_dev,verts_to_send_view);
+      Kokkos::deep_copy(boundary_verts_host, boundary_verts_dev);
+      gno_t boundary_verts_size = verts_to_send_size(0);
+      
+      //give the entire local graph to KokkosKernels to color
+      if(verbose){
+        std::cout<<comm->getRank()<<" coloring interior\n";
+        comm->barrier();
+      }
+      interior_time = timer();
+      total_time = timer();
+      if(n_local > 0){
+        this->colorInterior<execution_space, memory_space, memory_space>
+		(n_local, dist_adjs_dev, dist_offsets_dev, femv, verts_to_recolor_view, verts_to_recolor_size(0), false);
+      }
+      interior_time = timer() - interior_time;
+      comp_time = interior_time;
+      
+      Kokkos::View<int*, device_type> ghost_colors("ghost color backups", rand.size() - n_local);
```

Comment:
Add comments:  what does ghost colors store?  how is it indexed?  Why is it sized relative to rand.size()?